### DISCUSSION TITLE
Best practices for Sign-In with Ethereum (SIWE) implementation

### DISCUSSION THREAD

**Discussion by MidnightLightning** [wevm/wagmi#1989] [Q&A]
I have some questions about the suggested method the "Sign-In with Ethereum" (SIWE; [ERC-4361](https://eips.ethereum.org/EIPS/eip-4361)) process can be implemented (https://wagmi.sh/examples/sign-in-with-ethereum). The current suggestion is to use the `iron-session` library, which is a stateless session tool. However, being stateless on the server side I believe means it's not using nonces the way the EIP is intending.

Iron-session gives users encrypted blobs of data ("seals") that the browser is instructed to store as a cookie and then give back to the server on every request. Only the server that generated them knows how to decrypt the seal, and the data inside is the user's  session info. The way this example is laid out now, it does ensure that the login flow produces a seal that has a specific nonce and a signature for a SIWE-formatted message using that nonce. However, once the seal is in the user's hands, it could be copied from the user's cookie storage (the user cannot modify the encrypted blob, but they could copy it in its entirety). The "logout" endpoint as currently defined instructs the user's browser to destroy the seal it has in storage, but _**it doesn't invalidate the nonce used in that "logged-out" seal**_. So, even after the user has "logged out", if they (or an attacker) sets their own cookie manually to the backed-up seal value, they would be "logged back in again".

Notably, this means that if someone knows their session got hijacked (bad actor got a copy of their logged-in seal blob), there's nothing the user can do to invalidate the stolen session immediately. If the SIWE message had an "Expiration Time" defined, that would invalidate the attacker's session after that span. But that's an optional setting, and currently the example implementation doesn't use that. Iron-session also has a TTL/expiration setting, and that does invalidate the seal if the expiration date is tampered with or if that date has passed. So that's the only thing in this scenario that would stop an attacker from continuing to use that session.

An attacker being "logged in" in this scenario wouldn't allow them to do anything on-chain (which is good), but anything off-chain the application grants the user access to, the attacker would have access to.

The goal of stateless sessions is that each request from the client/user is as if it were the first request of their visit. The user provides all the data to "log in" with every request. But the server is still expected to check and see if what's supplied is valid on every request. For this setup, in addition to verifying the supplied signature matches the message, it should determine if the nonce is valid before accepting the request as authenticated.

Assuming all this logic is correct, I'd recommend updating the example to track which nonce was most-recently-generated for each address (the `/api/verify` endpoint, if it reaches the point to do a `req.session.save()`, should also record that nonce as now "active" for that address. The `/api/me` endpoint should check that the decrypted session data has the nonce that's active for the user, and the `/api/logout` endpoint should clear all nonces for that address). This can be a simple file cache to give as an example implementation that's not too complicated (doing a memory cache would be even simpler, but the downside of logging everyone out when the server resets is not likely a good feature for even the simplest of applications).

Or, more notice should be given on that example page illustrating that this style of nonce-usage does not prevent replay attacks, and encourage developers to use some sort of nonce-invalidation if their application grants significant off-chain benefits to logged-in users.

**Comment by tmm** [comment]
Hi there! The wagmi docs SIWE example definitely could use an update and is not a complete solution by any means. The SpruceID (creators of SIWE) examples seem like a good place to check out: https://github.com/spruceid/siwe

**Reply by MidnightLightning** [reply]
The examples in that repository do use a stateful session process (using `express-session` with `session-file-store`: https://github.com/spruceid/siwe-notepad/blob/main/src/index.ts#L14), or don't have a logout process at all (https://github.com/spruceid/siwe-quickstart/blob/main/02_backend/src/index.js), so neither of those cover Next.js.

I could definitely help contribute a draft of changing the `wagmi` docs, but want to first understand why they are the way they are, first.

**Reply by Garabed96** [reply]
Hey @MidnightLightning 
I'm not sure if I'm missing something here or its been updated. 
Going based off these wagmi [docs](https://1.x.wagmi.sh/examples/sign-in-with-ethereum)

How would an attacker modify the TTL of an existing session, can't the server only read or modify the session contents. Manipulating the seal by an attacker should invalidate the seal. Also isn't the purpose of the nonce to be a one-time use token to prevent this replay attack.
So the implication of this vulnerability would be such that once the data is stored in the clients cookie a third party could just steal that until TTL expiry if they haven't logged out?  

Edit: Ok seems like I misunderstood this, sessions can't be instantly invalidated, since by default no state is stored about session on the server. To disconnect a customer we would add a `isBlocked` state in the dB and create a UI to block customers. Would require checking the flag before reading / altering sensitive data.

Wondering at this point if I should just use an alternative for my nextjs app

**Comment by MidnightLightning** [comment]
@Garabed96 to clarify: the example in the current documentation lays out using `iron-session` which doesn't keep any state on the server-side. The example in the documentation doesn't modify the TTL of the session that Iron creates, so uses the default 14 days for the length of time the user will be "logged in" by the "seal" session cookie created by that example.

The issue I was highlighting in this discussion is if the user knows they got session-hijacked (an attacker copied their entire "seal" cookie. Didn't modify it, but copied the encrypted data blob from the victim's browser), they have no way to stop the attack. The example "logout" only destroys that user's copy of the "seal", not the one the attacker already copied. So however much of that 14 days are left for that session, the attacker can impersonate the victim.

The documentation has been updated with a notice that you'd want to implement some form of protection from session duplication. Which is true, but not not a full example of how to actually do that. If you'd like to see an example of how I accomplished it in NextJS (using a Firestore database as back-end for storing session info), you can take a peek [over here](https://gitlab.com/mooncatrescue/chainstation-web/-/tree/master/pages/api). In that implementation, users are able to "log in" as multiple different addresses, and so the `iron-session` data blob has a `keyring` which is an array of valid sessions, rather than a single valid session.
